### ISSUE TITLE
Issues creating nodegroup, Cryptic error message

### ISSUE THREAD

**Issue by andrewhertog** [pulumi/pulumi-eks#566]
## Expected behavior

Successfully create a nodegroup attached to a cluster

## Current behavior

Erroring with the following message:
```
    error: program failed: marshaling properties: awaiting input property cluster: missing one or more required arguments
    exit status 1

    error: an unhandled error occurred: program exited with non-zero exit code: 1
```
## Steps to reproduce

run `pulumi up` with the following code snippet:
```go

    cluster, err := eks.NewCluster(ctx, stack, &eks.ClusterArgs{
      SkipDefaultNodeGroup: pulumi.Bool(true),
      VpcId: pulumi.String(vpcId),
      SubnetIds: pulumi.StringArray{
        pulumi.String("subnet-<redacted>"),
        pulumi.String("subnet-<redacted>"),
        pulumi.String("subnet-<redacted>"),
      },
      ProviderCredentialOpts: &eks.KubeconfigOptionsArgs{
        ProfileName: pulumi.String(awsProfileName),
      },
      EndpointPrivateAccess: pulumi.Bool(true),
      EndpointPublicAccess: pulumi.Bool(false),
      ClusterSecurityGroup: clusterSecurityGroup,
      InstanceProfileName: instanceProfile.Name,
      NodeAssociatePublicIpAddress: pulumi.Bool(false),
    })
    if err != nil {
      return err
    }

_, err = eks.NewNodeGroup(ctx, fmt.Sprintf("%s-%s",stack,"systemNg") , &eks.NodeGroupArgs{
      Cluster: cluster.Core,
      InstanceProfile: instanceProfile,
      DesiredCapacity: pulumi.Int(1),
      MinSize: pulumi.Int(1),
      MaxSize: pulumi.Int(2),
      InstanceType: pulumi.String("t2.medium"),
      Labels: pulumi.StringMap{
        "ondemand": pulumi.String("true"),
        "nodegroup": pulumi.String("system"),
        "zone": pulumi.String("default"),
      },
      NodeSubnetIds: pulumi.StringArray{
        pulumi.String("subnet-<redacted>"), // same subnets as cluster
        pulumi.String("subnet-<redacted>"),
        pulumi.String("subnet-<redacted>"),
      },
    },pulumi.Parent(cluster))
```

## Context (Environment)

It created the cluster succesfully, but failed with the obscure message posted above

Running on Mac Catalina
Pulumi version 2.25.1 (Updated to 3.0.0 with the same results)

Iam resources and security groups created using pulumi-aws module

Full Preview and Output of `pulumi up`
```
❯ pulumi up
Previewing update (andrew-dev)

View Live: https://app.pulumi.com/andrewhertog/proto-pulumi/andrew-dev/previews/1aa49f04-967c-459d-964e-982b74d926b8

     Type                                   Name                              Plan
 +   pulumi:pulumi:Stack                    proto-pulumi-andrew-dev           create
 +   ├─ aws:iam:Role                        ec2InstanceRole                   create
 +   ├─ aws:ec2:SecurityGroup               andrew-dev-clusterSg              create
 +   ├─ aws:iam:RolePolicyAttachment        ec2InstanceRole-policy-2          create
 +   ├─ aws:iam:RolePolicyAttachment        ec2InstanceRole-policy-1              create
 +   ├─ aws:iam:RolePolicyAttachment        ec2InstanceRole-policy-0              create
 +   ├─ aws:iam:InstanceProfile             andrew-dev-instanceProfile                    create
 +   └─ eks:index:Cluster                   andrew-dev                                    create
 +      ├─ eks:index:ServiceRole            andrew-dev-instanceRole                       create
 +      │  ├─ aws:iam:Role                  andrew-dev-instanceRole-role                  create
 +      │  ├─ aws:iam:RolePolicyAttachment  andrew-dev-instanceRole-03516f97              create
 +      │  ├─ aws:iam:RolePolicyAttachment  andrew-dev-instanceRole-3eb088f2               create
 +      │  └─ aws:iam:RolePolicyAttachment  andrew-dev-instanceRole-e1b295bd               create
 +      ├─ eks:index:ServiceRole            andrew-dev-eksRole                                     create
 +      │  ├─ aws:iam:Role                  andrew-dev-eksRole-role                                create
 +      │  ├─ aws:iam:RolePolicyAttachment  andrew-dev-eksRole-90eb1c99                            create
 +      │  └─ aws:iam:RolePolicyAttachment  andrew-dev-eksRole-4b490823                            create
 +      ├─ aws:eks:Cluster                  andrew-dev-eksCluster                                  create
 +   pulumi:pulumi:Stack                    proto-pulumi-andrew-dev                                create
 +      ├─ pulumi:providers:kubernetes      andrew-dev-provider                                    create
 +      ├─ eks:index:VpcCni                 andrew-dev-vpc-cni                                     create
 +      ├─ aws:ec2:SecurityGroup            andrew-dev-nodeSecurityGroup                           create
 +      ├─ kubernetes:core/v1:ConfigMap     andrew-dev-nodeAccess                                  create
 +      ├─ aws:ec2:SecurityGroupRule        andrew-dev-eksNodeIngressRule                          create
 +      ├─ aws:ec2:SecurityGroupRule        andrew-dev-eksNodeClusterIngressRule                   create
 +      ├─ aws:ec2:SecurityGroupRule        andrew-dev-eksClusterIngressRule                       create
 +      ├─ aws:ec2:SecurityGroupRule        andrew-dev-eksExtApiServerClusterIngressRule           create
 +      ├─ aws:ec2:SecurityGroupRule        andrew-dev-eksNodeInternetEgressRule                   create
 +      └─ eks:index:NodeGroup              andrew-dev-systemNg                                    create
 +         ├─ eks:index:RandomSuffix        andrew-dev-systemNg-cfnStackName                       create
 +         ├─ aws:ec2:SecurityGroup         andrew-dev-systemNg-nodeSecurityGroup                  create
 +         ├─ aws:ec2:SecurityGroupRule     andrew-dev-systemNg-eksNodeInternetEgressRule          create
 +         ├─ aws:ec2:SecurityGroupRule     andrew-dev-systemNg-eksNodeIngressRule                 create
 +         ├─ aws:ec2:SecurityGroupRule     andrew-dev-systemNg-eksExtApiServerClusterIngressRule  create
 +         ├─ aws:ec2:SecurityGroupRule     andrew-dev-systemNg-eksNodeClusterIngressRule          create
 +         ├─ aws:ec2:SecurityGroupRule     andrew-dev-systemNg-eksClusterIngressRule              create
 +         ├─ aws:ec2:LaunchConfiguration   andrew-dev-systemNg-nodeLaunchConfiguration            create
 +         └─ aws:cloudformation:Stack      andrew-dev-systemNg-nodes                              create

Resources:
    + 38 to create

Do you want to perform this update? yes
Updating (andrew-dev)

View Live: https://app.pulumi.com/andrewhertog/proto-pulumi/andrew-dev/updates/15

     Type                                   Name                              Status       Info
 +   pulumi:pulumi:Stack                    proto-pulumi-andrew-dev           creating..
 +   pulumi:pulumi:Stack                    proto-pulumi-andrew-dev           creating...
 +   pulumi:pulumi:Stack                    proto-pulumi-andrew-dev           creating..
 +   ├─ aws:iam:RolePolicyAttachment        ec2InstanceRole-policy-0              created
 +   ├─ aws:iam:InstanceProfile             andrew-dev-instanceProfile                    created
 +   ├─ aws:iam:RolePolicyAttachment        ec2InstanceRole-policy-2                      created
 +   ├─ aws:iam:RolePolicyAttachment        ec2InstanceRole-policy-1                      created
 +   └─ eks:index:Cluster                   andrew-dev                                    created
 +      ├─ eks:index:ServiceRole            andrew-dev-instanceRole                       created
 +      │  ├─ aws:iam:Role                  andrew-dev-instanceRole-role                  created
 +      │  ├─ aws:iam:RolePolicyAttachment  andrew-dev-instanceRole-03516f97              created
 +      │  ├─ aws:iam:RolePolicyAttachment  andrew-dev-instanceRole-3eb088f2              created
 +      │  └─ aws:iam:RolePolicyAttachment  andrew-dev-instanceRole-e1b295bd              created
 +      ├─ eks:index:ServiceRole            andrew-dev-eksRole                            created
 +      │  ├─ aws:iam:Role                  andrew-dev-eksRole-role                       created
 +      │  ├─ aws:iam:RolePolicyAttachment  andrew-dev-eksRole-90eb1c99                   created
 +      │  └─ aws:iam:RolePolicyAttachment  andrew-dev-eksRole-4b490823                   created
 +      ├─ aws:eks:Cluster                  andrew-dev-eksCluster                         created
 +      ├─ aws:ec2:SecurityGroup            andrew-dev-nodeSecurityGroup                  created
 +      ├─ eks:index:VpcCni                 andrew-dev-vpc-cni                            created
 +      ├─ pulumi:providers:kubernetes      andrew-dev-provider                           created
 +      ├─ pulumi:providers:kubernetes      andrew-dev-eks-k8s                            created
 +      ├─ kubernetes:core/v1:ConfigMap     andrew-dev-nodeAccess                         created
 +      ├─ aws:ec2:SecurityGroupRule        andrew-dev-eksNodeInternetEgressRule          created
 +      ├─ aws:ec2:SecurityGroupRule        andrew-dev-eksExtApiServerClusterIngressRule  created
 +      ├─ aws:ec2:SecurityGroupRule        andrew-dev-eksClusterIngressRule              created
 +      ├─ aws:ec2:SecurityGroupRule        andrew-dev-eksNodeIngressRule                 created
 +      └─ aws:ec2:SecurityGroupRule        andrew-dev-eksNodeClusterIngressRule          created

Diagnostics:
  pulumi:pulumi:Stack (proto-pulumi-andrew-dev):
    error: program failed: marshaling properties: awaiting input property cluster: missing one or more required arguments
    exit status 1

    error: an unhandled error occurred: program exited with non-zero exit code: 1

Resources:
    + 28 created

Duration: 10m48s



```

## Affected feature

Golang eks.NewNodeGroup is suspected to have the issue.

**Comment by andrewhertog** [comment]
I no longer see this error anymore with `v0.30.0` of `pulumi-eks`, but now i get the following error with no code changes other than the latest pulumi modules
```
Diagnostics:
  pulumi:pulumi:Stack (proto-pulumi-andrew-dev):
    error: program failed: marshaling properties: awaiting input property cluster: cannot marshal an input of type pulumi.StringOutput with element type string as a value of type pulumi.StringOutput
    exit status 1

    error: an unhandled error occurred: program exited with non-zero exit code: 1

    error: [runtime] Error: Should not call getSessionAsync unless on Node11 or above.
        at Object.<anonymous> (/Users/andrew.denhertog/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/runtime/closure/v8Hooks.js:67:19)
        at Generator.next (<anonymous>)
        at /Users/andrew.denhertog/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/runtime/closure/v8Hooks.js:21:71
        at new Promise (<anonymous>)
        at __awaiter (/Users/andrew.denhertog/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/runtime/closure/v8Hooks.js:17:12)
        at Object.getSessionAsync (/Users/andrew.denhertog/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/runtime/closure/v8Hooks.js:65:12)
        at /Users/andrew.denhertog/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/runtime/closure/v8_v11andHigher.js:106:39
        at Generator.next (<anonymous>)
        at /Users/andrew.denhertog/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/runtime/closure/v8_v11andHigher.js:21:71
        at new Promise (<anonymous>)


```

**Comment by andrewhertog** [comment]
Updating my systems Node version to v14 gets rid of the getSessionAsync error, but now only have the marshaling error. I'm assuming it has to do with one of the inputs, but not sure which one

**Comment by universam1** [comment]
probably same problem here also facing a marshaling error
`    error: program failed: marshaling properties: awaiting input property cluster: cannot marshal an input of type pulumi.StringOutput with element type string as a value of type pulumi.StringOutput
    exit status 1`

Any progress or ideas in that regard @justinvp @andrewhertog ?
Thanks!

**Comment by andrewhertog** [comment]
Yes, looks like the same issue. I've given up on Pulumi for the time being tbh....

**Comment by universam1** [comment]
Yeah it looks like the main libs of Pulumi are broken at least regarding Golang - shows how business critical acceptance tests are. Too bad we would have loved to adapt it.

**Comment by justinvp** [comment]
Sorry for the trouble and thanks for your patience, @andrewhertog and @universam1. We'll take a closer look at this today and let you know.

**Comment by justinvp** [comment]
I've reproduced the error, but need to spend more time debugging to understand the root cause. May need to pick this back up next week.

**Comment by justinvp** [comment]
https://github.com/pulumi/pulumi/pull/7158 fixes the underlying Go marshaling issues. Once that's merged, we'll need to regenerate the Go SDK for EKS and release a new version.

That change is sufficient to get managed node groups working (i.e. using `eks.NewManagedNodeGroup`; see the example here: https://github.com/pulumi/pulumi-eks/issues/577#issuecomment-849879316).

However, using the older `eks.NewNodeGroup` yields a new error from inside the TypeScript implementation:

```
    error: Error: 'dependsOn' was passed a value that was not a Resource: 523e1e61a9377aa8
        at /Users/justin/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/runtime/resource.js:554:27
        at Generator.next (<anonymous>)
        at /Users/justin/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/runtime/resource.js:21:71
        at new Promise (<anonymous>)
        at __awaiter (/Users/justin/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/runtime/resource.js:17:12)
        at gatherExplicitDependencies (/Users/justin/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/runtime/resource.js:532:12)
        at /Users/justin/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/runtime/resource.js:546:50
        at /Users/justin/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/output.js:250:35
        at Generator.next (<anonymous>)
        at /Users/justin/.pulumi/plugins/resource-eks-v0.30.0/node_modules/@pulumi/pulumi/output.js:21:71
```

`cluster.Core` includes the `VpcCni` resource, which is marshaled as a resource reference, and is used by the `NodeGroup` component. The problem is we do not have a resource registration on the TypeScript side for the `VpcCni` resource, so when the resource reference is deserialized, instead of rehydrating an instance of `VpcCni` resource, the deserializer falls back to the ID string of the resource, which cannot be used with `dependsOn`. The fix will involve adding a resource registration for `VpcCni`.

**Comment by itay** [comment]
@justinvp is the fix going to be shipped anytime soon? It is pretty much a blocking issue for being able to use the library in Go, as without it you can't really create managed node groups.

**Comment by lukehoban** [comment]
> @justinvp is the fix going to be shipped anytime soon? 

Version 0.31.0 is now available at https://github.com/pulumi/pulumi-eks/releases/tag/v0.31.0 with the fix for `eks. NewManagedNodeGroup`.

The issue being tracked here for `eks.NewNodeGroup` is still under investigation.  https://github.com/pulumi/pulumi-eks/pull/590 addresses part of that.

**Comment by itay** [comment]
Thanks @lukehoban - I tried it with Pulumi 3.4.0 and the new pulumi-eks version and the issue is indeed fixed!

**Comment by justinvp** [comment]
#590 fixed the `dependOn` issue, which was released in `v0.32.0`.
